04-onion-routing.md

```diff
@@ -263,6 +264,12 @@ It is formatted according to the Type-Length-Value format defined in [BOLT #1](0
     2. data:
         * [`32*byte`:`payment_secret`]
         * [`tu64`:`total_msat`]
+    1. type: 10 (`encrypted_recipient_data`)
+    2. data:
+        * [`...*byte`:`encrypted_data`]
+    1. type: 12 (`blinding_point`)
+    2. data:
+        * [`point`:`blinding`]
```

Comment:
The blinding point should be provided by the previous node in the route as part of the `update_add_htlc`, the sender shouldn't even know the intermediate blinding points.

Reply:
That is true _inside_ the blinded route, but not _outside_ of it. If we want to support payments to blinded routes before the whole network supports route blinding, we have to provide the _first_ blinding point in the onion (note that it's only for the first blinding point).

A concrete example:

- Eve wants to receive a payment through the blinded route `Carol -> Dave -> Eve`
- Alice wants to pay this, but the only route it has to Carol is through Bob who doesn't support route blinding
- Alice uses the route `Alice -> Bob -> Carol -> (blinded) Dave -> (blinded) Eve` and puts Carol's blinding point in the onion for Carol

Carol will then use a tlv in `update_add_htlc` to provide the blinding point for Dave, and Dave will do the same to provide the blinding point for Eve.

Reply:
This is detailed here: https://github.com/lightning/bolts/blob/route-blinding/proposals/route-blinding.md#sending-to-a-blinded-route

It's probably worth a comment in the `Rationale` section here, I'll improve this.

Reply:
The way I understand it, [proposals/route-blinding.md](https://github.com/lightning/bolts/blob/route-blinding/proposals/route-blinding.md) is not part of the spec itself, you should write that in the `update_add_htlc` section of the spec.

Reply:
The reason it isn't there yet is that this PR currently doesn't specify how to use route blinding for payments, as we initially only needed it for onion messages (which doesn't require anything payment-related until offers is implemented).

I was waiting for more feedback on the open issues described in the `proposals` document, but you're right that I should publish a follow-up PR that builds on top of this one and specifies blinded payments. I have an old branch around, I'll work on this soon.

Reply:
> Is it critical or just "least privilege" that the sender only have knowledge of the first key in the chain of ephemeral keys, E(0) and not know all of the ephemeral keys, E(i>0)?

It's just "least privilege", even if the sender had all the ephemeral keys they wouldn't be able to derive any secret, that would require access to the relaying node's private keys (please do check the maths to make sure that statement is correct!).

Reply:
That's what I was thinking as well. The computation of the shared secrets between recipient and blinded routing nodes looks good to me. Thanks!

Reply:
That's an interesting point. We can change the format of the onion, that shouldn't be what's blocking us (the `1366` hard-coded value can be lifted whenever we want as long as it's gated by a feature bit - we can technically decide in the future that we want our onions to have 2500 bytes for example if we're running out of space). So what you suggest is completely doable, it's really just a matter of choosing where we think it's most convenient to have this data.

Since the `blinding_point` is necessary to decrypt the onion, it would make sense to directly include it in the onion structure as you suggest. If we plan on using this field for potentially other use-cases (no idea whether that will happen or not), it makes sense to have it in a separate tlv field. Both solutions work, it's probably a matter of choosing which one is simplest based on existing implementations. For `eclair` both would be fine, it's probably a bit more work to put this in the onion than in a separate tlv, but definitely doable. What about other implementations?

Reply:
Actually scratch that, I thought we had the `onion_routing_packet` in a tlv field in `update_add_htlc`, which would have let us define a new one and omit the old one. But it's actually hard-coded in the "core" of the message, so we can't do it easily (unless we define a new `update_add_htlc_v2` message, which is too cumbersome imho for just route blinding).

Reply:
cool, makes sense :+1: